File: packages/flutter_core/lib/core/services/sync/queue/queue_up.dart
import 'dart:async';
import '../../database/local.dart';
import '../../mutation/events.dart'; // MutationEvent, MutationPayload
import '../events.dart'; // SyncQueueEventSystem, SyncQueueEventTypes
import '../../events/main.dart'; // BaseEvent, EventSystem
import '../../query/utils.dart'; // extractResult

// Internal type for valid UpEvents used in the queue
abstract class UpEvent {
  final String mutationId;
  final String recordId;
  final Map<String, dynamic>? data;

  UpEvent(this.mutationId, this.recordId, {this.data});
}

class CreateEvent extends UpEvent {
  CreateEvent(String mutationId, String recordId, Map<String, dynamic> data)
    : super(mutationId, recordId, data: data);
}

class UpdateEvent extends UpEvent {
  UpdateEvent(String mutationId, String recordId, Map<String, dynamic> data)
    : super(mutationId, recordId, data: data);
}

class DeleteEvent extends UpEvent {
  DeleteEvent(String mutationId, String recordId) : super(mutationId, recordId);
}

class UpQueue {
  final List<UpEvent> queue = [];
  final EventSystem<SyncQueueEvent> events = EventSystem<SyncQueueEvent>();

  final LocalDatabaseService local;

  UpQueue(this.local);

  int get size => queue.length;

  void push(UpEvent event) {
    queue.add(event);
    events.addEvent(
      MutationEnqueued(MutationEnqueuedPayload(queueSize: queue.length)),
    );
  }

  Future<void> next(Future<void> Function(UpEvent event) fn) async {
    if (queue.isEmpty) return;

    final event = queue.removeAt(0);

    try {
      await fn(event);
    } catch (error) {
      print("Failed to process mutation: $error");
      // Put it back at the front
      queue.insert(0, event);
      rethrow;
    }

    try {
      await removeEventFromDatabase(event.mutationId);
    } catch (error) {
      print("Failed to remove mutation from database: $error");
    }
  }

  Future<void> removeEventFromDatabase(String mutationId) {
    return local.query("DELETE type::record(\$id)", vars: {'id': mutationId});
  }

  void listenForMutations(EventSystem<dynamic> mutationEvents) {
    // Subscribe to MutationEvent
    mutationEvents.subscribe<MutationEvent>((event) {
      for (final mutation in event.payload) {
        _handleMutationPayload(mutation);
      }
    });
  }

  void _handleMutationPayload(MutationPayload mutation) {
    final event = _payloadToUpEvent(mutation);
    if (event != null) push(event);
  }

  void _addToQueueInternal(MutationPayload mutation) {
    final event = _payloadToUpEvent(mutation);
    if (event != null) queue.add(event);
  }

  UpEvent? _payloadToUpEvent(MutationPayload mutation) {
    switch (mutation.type) {
      case MutationAction.create:
        if (mutation.data != null) {
          return CreateEvent(
            mutation.mutation_id,
            mutation.record_id,
            mutation.data!,
          );
        }
        break;
      case MutationAction.update:
        if (mutation.data != null) {
          return UpdateEvent(
            mutation.mutation_id,
            mutation.record_id,
            mutation.data!,
          );
        }
        break;
      case MutationAction.delete:
        return DeleteEvent(mutation.mutation_id, mutation.record_id);
    }
    return null;
  }

  Future<void> loadFromDatabase() async {
    try {
      final resStr = await local.query<String>(
        "SELECT * FROM _spooky_pending_mutations ORDER BY created_at ASC",
      );

      final dynamic raw = extractResult(resStr);
      List<Map<String, dynamic>> mutations = [];

      if (raw is List) {
        mutations = raw.cast<Map<String, dynamic>>();
      }

      // Match TS: Clear queue/cache from DB as authoritative source
      queue.clear();

      for (final r in mutations) {
        final typeStr = r['mutationType'] as String?;
        final recordId = r['recordId'] as String?;
        final data = r['data'] as Map<String, dynamic>?;
        final mutationId = r['id'].toString();

        if (typeStr == null || recordId == null) continue;

        MutationAction? action;
        try {
          action = MutationAction.values.firstWhere((e) => e.name == typeStr);
        } catch (_) {}

        if (action == null) continue;

        final payload = MutationPayload(
          type: action,
          record_id: recordId,
          mutation_id: mutationId,
          data: data,
        );

        // Directly add to queue without emitting events (TS behavior)
        _addToQueueInternal(payload);
      }
    } catch (error) {
      print("Failed to load pending mutations: $error");
    }
  }
}

File: packages/core/src/services/sync/queue/queue-up.ts
import { RecordId } from 'surrealdb';
import { MutationEventSystem, MutationEventTypes } from '../../mutation/events.js';
import { LocalDatabaseService } from '../../database/index.js';
import {
  createSyncQueueEventSystem,
  SyncQueueEventSystem,
  SyncQueueEventTypes,
} from '../events.js';

export type CreateEvent = {
  type: 'create';
  mutation_id: RecordId;
  record_id: RecordId;
  data: Record<string, unknown>;
};

export type UpdateEvent = {
  type: 'update';
  mutation_id: RecordId;
  record_id: RecordId;
  data: Record<string, unknown>;
};

export type DeleteEvent = {
  type: 'delete';
  mutation_id: RecordId;
  record_id: RecordId;
};

export type UpEvent = CreateEvent | UpdateEvent | DeleteEvent;

export class UpQueue {
  private queue: UpEvent[] = [];
  private _events: SyncQueueEventSystem;

  get events(): SyncQueueEventSystem {
    return this._events;
  }

  constructor(private local: LocalDatabaseService) {
    this._events = createSyncQueueEventSystem();
  }

  get size(): number {
    return this.queue.length;
  }

  push(event: UpEvent) {
    this.queue.push(event);
    this._events.addEvent({
      type: SyncQueueEventTypes.MutationEnqueued,
      payload: { queueSize: this.queue.length },
    });
  }

  async next(fn: (event: UpEvent) => Promise<void>): Promise<void> {
    const event = this.queue.shift();
    if (event) {
      try {
        await fn(event);
      } catch (error) {
        console.error('Failed to process mutation', event, error);
        this.queue.unshift(event);
        throw error;
      }
      try {
        await this.removeEventFromDatabase(event.mutation_id);
      } catch (error) {
        // TODO: handle this, we still have this mutation in the database, eventough it
        // was processed successfully
        console.error('Failed to remove mutation from database', event, error);
      }
    }
  }

  async removeEventFromDatabase(mutation_id: RecordId) {
    return this.local.query(`DELETE $mutation_id`, { mutation_id });
  }

  async loadFromDatabase() {
    try {
      const rawResult = await this.local.query<any>(
        `SELECT * FROM _spooky_pending_mutations ORDER BY created_at ASC`
      );

      let records: any[] = [];

      // Handle potentially wrapped SurrealDB response
      if (Array.isArray(rawResult)) {
        records = rawResult;
      } else if (rawResult && typeof rawResult === 'object' && Array.isArray(rawResult.result)) {
        records = rawResult.result;
      }

      this.queue = records
        .map((r: any) => {
          switch (r.mutationType) {
            case 'create':
              return { type: 'create', mutation_id: r.id, record_id: r.recordId, data: r.data };
            case 'update':
              return { type: 'update', mutation_id: r.id, record_id: r.recordId, data: r.data };
            case 'delete':
              return { type: 'delete', mutation_id: r.id, record_id: r.recordId };
            default:
              console.warn(`Unknown mutation type: ${r.mutationType}`, r);
              return null;
          }
        })
        .filter((e): e is UpEvent => e !== null);
    } catch (error) {
      console.error('Failed to load pending mutations from database:', error);
      // TODO: clarify if we want to throw or not
      // Don't crash, just start with empty queue? Or throw?
      // For now, logging is safer.
    }
  }
}
